File: src-tauri/tests/work_graph.rs
// Phase 2 — Work Graph & core workflow. The acceptance scenario (company →
// contact → opportunity → proposal → won → project → kickoff meeting → meeting
// note → action item task → completed → Company 360), saved the way the app
// saves it: ids carried from the record the work was started from. Fictional
// names only.
use menabig_tracker_lib::activity::{query_activity, rename_company_row, ActivityFilter};
use menabig_tracker_lib::commands::{read_all_data, upsert_contact_rows, upsert_note_rows, upsert_proposal_rows, upsert_todo_rows};
use menabig_tracker_lib::db::init_connection;
use menabig_tracker_lib::integrity::integrity_report;
use menabig_tracker_lib::models::{Contact, Note, Proposal, Todo};
use menabig_tracker_lib::opportunities::{create_company_named, save_opportunity_row};
use menabig_tracker_lib::v2_commands::{save_meeting_row, save_project_row};
use menabig_tracker_lib::v2_models::{Meeting, Opportunity, Project};
use rusqlite::{params, Connection};

fn fresh_db(tag: &str) -> (std::path::PathBuf, Connection) {
    let path = std::env::temp_dir().join(format!("menabig_work_graph_{tag}_{}.sqlite3", std::process::id()));
    let _ = std::fs::remove_file(&path);
    let conn = init_connection(&path).expect("init db");
    (path, conn)
}

fn one<T: rusqlite::types::FromSql>(conn: &Connection, sql: &str) -> T {
    conn.query_row(sql, [], |r| r.get(0)).unwrap()
}

fn company_id(conn: &Connection, table: &str, id: i64) -> Option<i64> {
    conn.query_row(&format!("SELECT company_id FROM {table} WHERE id = ?1"), params![id], |r| r.get(0)).unwrap()
}

fn link(conn: &Connection, from: (&str, i64), to: (&str, i64)) {
    conn.execute(
        "INSERT OR IGNORE INTO entity_links (from_type, from_id, to_type, to_id, created_at) VALUES (?1, ?2, ?3, ?4, '2026-09-15')",
        params![from.0, from.1, to.0, to.1],
    )
    .unwrap();
}

struct Scenario {
    company: i64,
    rival: i64,
    contact: i64,
    opportunity: i64,
    proposal: i64,
    project: i64,
    meeting: i64,
    note: i64,
    task: i64,
}

/// Runs the acceptance scenario. A second company whose name differs only by
/// capitals exists, so anything resolved by name instead of id would be
/// ambiguous and show up as a wrong or missing company.
fn run_scenario(conn: &mut Connection) -> Scenario {
    // 1. Company.
    let company = create_company_named(conn, "Contoso Logistics").unwrap().unwrap();
    conn.execute("INSERT INTO companies (name, created_at) VALUES ('CONTOSO LOGISTICS', '2026-09-15')", []).unwrap();
    let rival: i64 = one(conn, "SELECT id FROM companies WHERE name = 'CONTOSO LOGISTICS'");
    // The company page's name, typed differently by someone: must still be the page's company.
    let typed = "contoso logistics";

    // 2. Contact, created from the company page.
    upsert_contact_rows(conn, &[Contact { id: 1, name: Some("Dana Test".into()), client_name: Some(typed.into()), company_id: Some(company), ..Default::default() }]).unwrap();

    // 3. Opportunity from the company page.
    let opp = save_opportunity_row(conn, &Opportunity {
        name: "Contoso Saudization".into(), company_id: Some(company), company_name: Some(typed.into()), stage: "Lead".into(), status: "Open".into(), ..Default::default()
    }).unwrap();
    link(conn, ("contact", 1), ("opportunity", opp.id));

    // 4. Proposal from the opportunity (inherits company; linked to the opportunity).
    upsert_proposal_rows(conn, &[Proposal { id: 1, client: typed.into(), company_id: Some(company), status: "Drafting".into(), r#type: Some("Saudization".into()), ..Default::default() }]).unwrap();
    let opp = save_opportunity_row(conn, &Opportunity { proposal_id: Some(1), stage: "Proposal".into(), ..opp }).unwrap();

    // 5. Won.
    let opp = save_opportunity_row(conn, &Opportunity { stage: "Won".into(), ..opp }).unwrap();

    // 6. Project from the won opportunity ("Create Project" is an explicit action).
    let project = save_project_row(conn, &Project {
        name: "Contoso Saudization Project".into(), r#type: "client".into(), status: "Planning".into(), priority: "High".into(),
        company_name: Some(typed.into()), company_id: Some(company), ..Default::default()
    }).unwrap();
    let opp = save_opportunity_row(conn, &Opportunity { project_id: Some(project.id), ..opp }).unwrap();

    // 7. Kickoff meeting from the project: company and project.
    let meeting = save_meeting_row(conn, &Meeting {
        title: "Contoso Saudization Kickoff".into(), meeting_date: Some("2026-09-15".into()),
        company_name: Some(typed.into()), company_id: Some(company), project_id: Some(project.id), ..Default::default()
    }).unwrap();

    // 8. Meeting note: company, linked to the project; the meeting points at it.
    upsert_note_rows(conn, &[Note {
        id: 1, title: Some("Contoso Saudization Kickoff — 15 Sep 2026".into()), content: Some("## Action Items\n- [ ] Send revised Saudization model".into()),
        client_name: Some(typed.into()), company_id: Some(company), ..Default::default()
    }]).unwrap();
    link(conn, ("note", 1), ("project", project.id));
    let meeting = save_meeting_row(conn, &Meeting { note_id: Some(1), ..meeting }).unwrap();

    // 9. The action item becomes a task: company, project and meeting; linked to the note.
    upsert_todo_rows(conn, &[Todo {
        id: 1, title: "Send revised Saudization model".into(), r#type: Some("client".into()), client: Some(typed.into()), company_id: Some(company),
        project_id: Some(project.id), meeting_id: Some(meeting.id), status: Some("Pending".into()), priority: Some("Medium".into()), ..Default::default()
    }]).unwrap();
    link(conn, ("task", 1), ("note", 1));

    // 10. Complete it.
    let mut task = read_all_data(conn).unwrap().todos.into_iter().find(|t| t.id == 1).unwrap();
    task.status = Some("Done".into());
    task.completed_at = Some("2026-09-15".into());
    upsert_todo_rows(conn, &[task]).unwrap();

    Scenario { company, rival, contact: 1, opportunity: opp.id, proposal: 1, project: project.id, meeting: meeting.id, note: 1, task: 1 }
}

#[test]
fn acceptance_scenario_links_every_record_by_id_and_persists() {
    let (path, mut conn) = fresh_db("scenario");
    let s = run_scenario(&mut conn);

    // Every record belongs to the company by id — never to the look-alike.
    for (table, id) in [("contacts", s.contact), ("opportunities", s.opportunity), ("proposals", s.proposal), ("projects", s.project), ("meetings", s.meeting), ("notes", s.note), ("todos", s.task)] {
        assert_eq!(company_id(&conn, table, id), Some(s.company), "{table} keeps the company it was created from");
    }
    for table in ["contacts", "opportunities", "proposals", "projects", "meetings", "notes", "todos"] {
        assert_eq!(one::<i64>(&conn, &format!("SELECT COUNT(*) FROM {table} WHERE company_id = {}", s.rival)), 0, "nothing moved to the look-alike company");
    }
    assert_eq!(one::<i64>(&conn, "SELECT COUNT(*) FROM company_review_queue"), 0, "ids avoided every ambiguity");

    // Reopen: the whole chain is still there.
    drop(conn);
    let conn = init_connection(&path).unwrap();
    let data = read_all_data(&conn).unwrap();
    assert_eq!(one::<Option<i64>>(&conn, &format!("SELECT proposal_id FROM opportunities WHERE id = {}", s.opportunity)), Some(s.proposal));
    assert_eq!(one::<Option<i64>>(&conn, &format!("SELECT project_id FROM opportunities WHERE id = {}", s.opportunity)), Some(s.project));
    assert_eq!(one::<String>(&conn, &format!("SELECT stage FROM opportunities WHERE id = {}", s.opportunity)), "Won");
    let meeting: (Option<i64>, Option<i64>, Option<i64>) = conn
        .query_row("SELECT project_id, note_id, company_id FROM meetings WHERE id = ?1", params![s.meeting], |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)))
        .unwrap();
    assert_eq!(meeting, (Some(s.project), Some(s.note), Some(s.company)));
    let task = data.todos.iter().find(|t| t.id == s.task).unwrap();
    assert_eq!((task.project_id, task.meeting_id, task.company_id, task.status.as_deref()), (Some(s.project), Some(s.meeting), Some(s.company), Some("Done")));
    assert_eq!(one::<i64>(&conn, &format!("SELECT COUNT(*) FROM entity_links WHERE from_type = 'task' AND from_id = {} AND to_type = 'note' AND to_id = {}", s.task, s.note)), 1);
    assert_eq!(one::<i64>(&conn, &format!("SELECT COUNT(*) FROM entity_links WHERE from_type = 'note' AND to_type = 'project' AND to_id = {}", s.project)), 1);

    // Company 360: every record is found by the company id.
    for (sql, expected) in [
        (format!("SELECT COUNT(*) FROM contacts WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM opportunities WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM proposals WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM projects WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM meetings WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM notes WHERE company_id = {}", s.company), 1),
        (format!("SELECT COUNT(*) FROM todos WHERE company_id = {} AND status = 'Done'", s.company), 1),
    ] {
        assert_eq!(one::<i64>(&conn, &sql), expected, "{sql}");
    }
    // Its activity tells the story once per event.
    let activity = query_activity(&conn, &ActivityFilter { company_id: Some(s.company), ..Default::default() }).unwrap();
    let has = |action: &str, entity: &str| activity.iter().filter(|a| a.action == action && a.entity_type == entity).count();
    assert_eq!(has("created", "meeting"), 1);
    assert_eq!(has("created", "note"), 1);
    assert_eq!(has("created", "task"), 1);
    assert_eq!(has("completed", "task"), 1);
    assert!(activity.iter().any(|a| a.entity_type == "opportunity"), "opportunity events: {activity:?}");
    assert_eq!(integrity_report(&conn).unwrap().count("activity recorded twice"), 0);
    let report = integrity_report(&conn).unwrap();
    assert_eq!(report.count("tasks whose opportunity belongs to another company"), 0);
    drop(conn);
    let _ = std::fs::remove_file(path);
}

#[test]
fn tasks_belong_to_opportunities_and_mismatches_are_reported_not_fixed() {
    let (path, mut conn) = fresh_db("task-opportunity");
    let s = run_scenario(&mut conn);
    let other = create_company_named(&conn, "Fabrikam").unwrap().unwrap();

    // A task created from the opportunity carries its id and shows in its activity.
    upsert_todo_rows(&mut conn, &[Todo { id: 2, title: "Chase signature".into(), client: Some("Contoso Logistics".into()), company_id: Some(s.company), opportunity_id: Some(s.opportunity), ..Default::default() }]).unwrap();
    let task = read_all_data(&conn).unwrap().todos.into_iter().find(|t| t.id == 2).unwrap();
    assert_eq!((task.opportunity_id, task.company_id), (Some(s.opportunity), Some(s.company)));
    assert_eq!(one::<Option<i64>>(&conn, "SELECT opportunity_id FROM activity WHERE entity_type = 'task' AND entity_id = 2"), Some(s.opportunity));

    // Explicit reassignment: choosing another company moves the task; its opportunity is left alone and the mismatch reported.
    upsert_todo_rows(&mut conn, &[Todo { client: Some("Fabrikam".into()), company_id: None, ..task.clone() }]).unwrap();
    assert_eq!(company_id(&conn, "todos", 2), Some(other));
    assert_eq!(one::<Option<i64>>(&conn, "SELECT opportunity_id FROM todos WHERE id = 2"), Some(s.opportunity));
    assert_eq!(integrity_report(&conn).unwrap().count("tasks whose opportunity belongs to another company"), 1);

    // No silent reassignment: saving unrelated fields keeps the company, even with stale text.
    let moved = read_all_data(&conn).unwrap().todos.into_iter().find(|t| t.id == 2).unwrap();
    upsert_todo_rows(&mut conn, &[Todo { description: Some("edited".into()), company_id: Some(s.company), ..moved }]).unwrap();
    assert_eq!(company_id(&conn, "todos", 2), Some(other), "a stale company id never undoes an explicit reassignment");

    // Renaming the company changes nothing about the links.
    rename_company_row(&conn, s.company, "Contoso Logistics KSA").unwrap();
    for (table, id) in [("opportunities", s.opportunity), ("projects", s.project), ("meetings", s.meeting), ("notes", s.note), ("todos", s.task)] {
        assert_eq!(company_id(&conn, table, id), Some(s.company));
    }

    // Deleting the opportunity unlinks (not deletes) its tasks.
    conn.execute("DELETE FROM opportunities WHERE id = ?1", params![s.opportunity]).unwrap();
    assert_eq!(one::<Option<i64>>(&conn, "SELECT opportunity_id FROM todos WHERE id = 2"), None);
    assert_eq!(one::<i64>(&conn, "SELECT COUNT(*) FROM todos"), 2);
    drop(conn);
    let _ = std::fs::remove_file(path);
}

File: src/lib/workGraph.ts
// The Work Graph: how records inherit context from the record they were
// created from, and which related records a record shows. Relationships are
// ids (company_id, project_id, opportunity_id, meeting_id and entity_links);
// names are only for display and are always read from the linked record.
//
// Context is a default, not a lock: the creation dialogs are prefilled from it
// and the person can change any field before saving.

import type { Agreement, Company, Contact, EntityKind, EntityLink, Meeting, Note, Opportunity, Project, Proposal, Todo } from './types';

/** The records a new record belongs to. */
export interface WorkContext {
  companyId: number | null;
  /** The company's current name (display and the text column), never used to identify it. */
  companyName: string | null;
  projectId: number | null;
  opportunityId: number | null;
  meetingId: number | null;
  noteId: number | null;
}

/** The slice of app state the helpers read. */
export interface GraphData {
  companies: Company[];
  opportunities: Opportunity[];
  projects: Project[];
  meetings: Meeting[];
  proposals: Proposal[];
  agreements: Agreement[];
  contacts: Contact[];
  todos: Todo[];
}

export const EMPTY_CONTEXT: WorkContext = { companyId: null, companyName: null, projectId: null, opportunityId: null, meetingId: null, noteId: null };

/** A record's company as { id, current name }. A linked company shows under its
 * current name; an unlinked record keeps its typed name with no id. */
export function companyOf(g: Pick<GraphData, 'companies'>, companyId: number | null | undefined, typedName: string | null | undefined): { companyId: number | null; companyName: string | null } {
  if (companyId != null) {
    const co = g.companies.find((c) => c.id === companyId);
    if (co) return { companyId: co.id, companyName: co.name };
  }
  const name = typedName?.trim() || null;
  return { companyId: null, companyName: name };
}

export function contextFromCompany(company: Company): WorkContext {
  return { ...EMPTY_CONTEXT, companyId: company.id, companyName: company.name };
}

export function contextFromOpportunity(g: GraphData, o: Opportunity): WorkContext {
  return { ...EMPTY_CONTEXT, ...companyOf(g, o.companyId, o.companyName), opportunityId: o.id };
}

/** A project's context: its company. The opportunity it was started from stays
 * reachable through the project (Origin), so project work isn't also filed
 * under the closed opportunity. */
export function contextFromProject(g: GraphData, p: Project): WorkContext {
  return { ...EMPTY_CONTEXT, ...companyOf(g, p.companyId, p.companyName), projectId: p.id };
}

export function contextFromMeeting(g: GraphData, m: Meeting): WorkContext {
  return { ...EMPTY_CONTEXT, ...companyOf(g, m.companyId, m.companyName), projectId: m.projectId, opportunityId: m.opportunityId, meetingId: m.id };
}

/** A note's context: its company, the project and opportunity it is linked to,
 * and the meeting whose notes it holds. `links` are the note's entity links. */
export function contextFromNote(g: GraphData, n: Note, links: EntityLink[]): WorkContext {
  const meeting = g.meetings.find((m) => m.noteId === n.id);
  const linked = (kind: EntityKind) => links.find((l) => l.fromType === 'note' && l.fromId === n.id && l.toType === kind)?.toId ?? null;
  const company = companyOf(g, n.companyId, n.clientName);
  return {
    ...EMPTY_CONTEXT,
    // A meeting note that has no company of its own takes the meeting's.
    ...(company.companyName || !meeting ? company : companyOf(g, meeting.companyId, meeting.companyName)),
    projectId: linked('project') ?? meeting?.projectId ?? null,
    opportunityId: linked('opportunity') ?? meeting?.opportunityId ?? null,
    meetingId: meeting?.id ?? null,
    noteId: n.id,
  };
}

/** A context with no company takes the company of its project, or else of its
 * opportunity. A company that is set is never replaced. */
export function inheritCompany(g: GraphData, ctx: WorkContext): WorkContext {
  if (ctx.companyName) return ctx;
  const p = ctx.projectId != null ? g.projects.find((x) => x.id === ctx.projectId) : undefined;
  const o = ctx.opportunityId != null ? g.opportunities.find((x) => x.id === ctx.opportunityId) : undefined;
  const from = p ? companyOf(g, p.companyId, p.companyName) : o ? companyOf(g, o.companyId, o.companyName) : null;
  return from?.companyName ? { ...ctx, ...from } : ctx;
}

/** Task fields for a task created in a context. */
export function taskFields(ctx: WorkContext): Partial<Todo> {
  return {
    type: ctx.companyName ? 'client' : 'general',
    client: ctx.companyName,
    companyId: ctx.companyName ? ctx.companyId : null,
    projectId: ctx.projectId,
    opportunityId: ctx.opportunityId,
    meetingId: ctx.meetingId,
  };
}

/** Company fields for a record saved from a form: the context's company id is
 * kept only while the company field still shows that company's name. Typing
 * another name is an explicit reassignment and is resolved by the backend. */
export function companyFromForm(ctx: Pick<WorkContext, 'companyId' | 'companyName'> | null, typed: string): { companyName: string | null; companyId: number | null } {
  const name = typed.trim() || null;
  if (!name) return { companyName: null, companyId: null };
  const same = ctx?.companyName != null && ctx.companyName.trim().toLowerCase() === name.toLowerCase();
  return { companyName: name, companyId: same ? ctx!.companyId : null };
}

// ── Links ────────────────────────────────────────────────────────────────────

/** A record's outgoing links with every link to `kind` replaced by `toIds`,
 * keeping links to other kinds (a note's project link must not drop its
 * opportunity link). */
export function replaceLinks(links: EntityLink[], fromType: EntityKind, fromId: number, kind: EntityKind, toIds: number[]): EntityLink[] {
  const others = links.filter((l) => l.fromType === fromType && l.fromId === fromId && l.toType !== kind);
  return [...others, ...[...new Set(toIds)].map((toId) => ({ fromType, fromId, toType: kind, toId }))];
}

/** Outgoing links plus the given ones (deduplicated). */
export function addLinks(links: EntityLink[], fromType: EntityKind, fromId: number, add: { toType: EntityKind; toId: number | null | undefined }[]): EntityLink[] {
  const out = links.filter((l) => l.fromType === fromType && l.fromId === fromId);
  for (const a of add) {
    if (a.toId == null) continue;
    if (!out.some((l) => l.toType === a.toType && l.toId === a.toId)) out.push({ fromType, fromId, toType: a.toType, toId: a.toId });
  }
  return out;
}

// ── Action items in notes ────────────────────────────────────────────────────

/** Unchecked Markdown checklist items ("- [ ] Send the model"), in order, without duplicates. */
export function actionItems(markdown: string | null | undefined): string[] {
  const out: string[] = [];
  for (const line of (markdown || '').split('\n')) {
    const m = /^\s*[-*+]\s+\[ \]\s+(.+?)\s*$/.exec(line);
    if (!m) continue;
    const title = m[1].replace(/\s+/g, ' ').trim();
    // "- [ ] >> …" is a commitment: it gets its task that way, not twice.
    if (/^(>>|<<)/.test(title)) continue;
    if (title && !out.some((t) => sameTitle(t, title))) out.push(title);
  }
  return out;
}

export function sameTitle(a: string, b: string): boolean {
  return a.trim().replace(/\s+/g, ' ').toLowerCase() === b.trim().replace(/\s+/g, ' ').toLowerCase();
}

/** Action items not yet turned into a task: a task linked to the note, or
 * created from its meeting, with the same title counts as done already. */
export function unconvertedActionItems(markdown: string | null | undefined, existing: Todo[]): string[] {
  return actionItems(markdown).filter((title) => !existing.some((t) => sameTitle(t.title, title)));
}

// ── Related records ──────────────────────────────────────────────────────────

/** The opportunity a project was started from. */
export function originOpportunity(g: Pick<GraphData, 'opportunities'>, projectId: number): Opportunity | undefined {
  return g.opportunities.find((o) => o.projectId === projectId);
}

/** A project's engagement chain: opportunity → proposal → agreement, and the
 * opportunity's contacts. `contactLinks` are the opportunity's entity links. */
export function projectChain(g: GraphData, projectId: number, contactLinks: EntityLink[] = []): { opportunity?: Opportunity; proposal?: Proposal; agreement?: Agreement; contacts: Contact[] } {
  const opportunity = originOpportunity(g, projectId);
  const proposal = opportunity?.proposalId != null ? g.proposals.find((p) => p.id === opportunity.proposalId) : undefined;
  const agreement = proposal ? g.agreements.find((a) => a.proposalId === proposal.id) : undefined;
  const contactIds = new Set(contactLinks.filter((l) => l.fromType === 'contact' && l.toType === 'opportunity' && l.toId === opportunity?.id).map((l) => l.fromId));
  return { opportunity, proposal, agreement, contacts: g.contacts.filter((c) => contactIds.has(c.id)) };
}

/** The project a proposal led to (through its opportunity). */
export function proposalProject(g: Pick<GraphData, 'opportunities' | 'projects'>, proposalId: number): Project | undefined {
  const o = g.opportunities.find((x) => x.proposalId === proposalId && x.projectId != null);
  return o ? g.projects.find((p) => p.id === o.projectId) : undefined;
}

/** Tasks for an opportunity: linked directly, or from one of its meetings. */
export function opportunityTasks(g: Pick<GraphData, 'todos' | 'meetings'>, opportunityId: number): Todo[] {
  const meetingIds = new Set(g.meetings.filter((m) => m.opportunityId === opportunityId).map((m) => m.id));
  return g.todos.filter((t) => t.opportunityId === opportunityId || (t.meetingId != null && meetingIds.has(t.meetingId)));
}
